safekeeper/src/receive_wal.rs

```diff
@@ -596,6 +628,10 @@ impl WalAcceptor {
                     WAL_RECEIVER_QUEUE_DEPTH.observe(self.msg_rx.len() as f64);
                     None // no reply
                 }
+
+                _ = self.tli.cancel.cancelled() => {
```

Comment:
Maybe write here that `process_msg()` is not cancellation-safe, which is why we're breaking here and letting any in-flight futures run to completion.

Reply:
Looking again at this function: because we already do the self.tli.process_msg inside a select!, it might be dropped anyway -- i.e. when we hit this cancellation token we might drop out the loop and leave some abandoned I/O in flight.

This is unsafe in general, but not immediately harmful because we currently only shut down timelines when they're being deleted.  We need to fix this before implementing offloading of inactive timelines.

In 2915e232a I add a comment linking to an issue https://github.com/neondatabase/neon/issues/9788

Reply:
Good catch -- I wasn't aware of this hazard when I made WalAcceptor event-driven, and even missed it while reviewing this change.

Having written a bunch of Go lately, I intuitively consider us having taken the select branch once we're in it, but I suppose with async the control flow is inverted and we only actually take the branch once we return from it. That messes with my head a little bit -- I'll need to read up on the subtleties here and internalize it.

Reply:
Wait, no, this isn't true. The futures we're selecting on are all cancellation safe:

* `self.msg_rx.recv()`
* `flush_ticker.tick()`
* `future::ready(()), if dirty && self.msg_rx.is_empty()`
* `metrics_ticker.tick()`

Once either of these futures are ready, we'll take the branch -- what happens inside the branch (i.e. `process_msg`) is no longer affected by the other branches of the `select`. Right?

Cognitive dissonance averted. :)

Reply:
> Wait, no, this isn't true.

+1. AFAIS we do process_msg only inside select! arm body, not condition.